### PR TITLE
IGNITE-20740 Fix SocketChannel leak when connection fails

### PR DESCRIPTION
**Pull request by shoothzj**
https://issues.apache.org/jira/browse/IGNITE-20740

### CODE REVIEW

**Comment by shoothzj** [comment]
@ptupitsyn @petrov-mg @ivan-fedorenkov @sk0x50  PTAL, thanks

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/io/gridnioserver/GridNioClientConnectionMultiplexer.java [hunk 1]
```diff
@@ -146,8 +146,30 @@ public GridNioClientConnectionMultiplexer(ClientConfiguration cfg) {
...
+                            throw new ClientConnectionException(ex.getMessage(), ex);
+                        }
+                    }
+                    try {
+                        ch.close();
+                    }
+                    catch (Exception ex) {
+                        throw new ClientConnectionException(ex.getMessage(), ex);
+                    }
+                }
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/io/gridnioserver/GridNioClientConnectionMultiplexer.java:170]:
```suggestion
                if (ch != null) {
                    if (ch.socket() != null) {
                        try {
                            ch.socket().close();
                        }
                        catch (Exception ignored) {
                        }
                    }
                    
                    try {
                        ch.close();
                    }
                    catch (Exception ignored) {
                    }
                }
```

The idea is to throw the original exception to the user. We don't care about exceptions that happen int `close()`.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/io/gridnioserver/GridNioClientConnectionMultiplexer.java (hunk 1 above)
Comment (shoothzj) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/io/gridnioserver/GridNioClientConnectionMultiplexer.java:170]: I agree. Can we do some log here?

**Comment by shoothzj** [comment]
@ptupitsyn Thanks for your review, Could you please review again? Thanks.

**Review by ptupitsyn — APPROVED** [review]

**Comment by ptupitsyn** [comment]
TC build triggered: https://ci.ignite.apache.org/viewLog.html?buildId=7591435&

**Comment by shoothzj** [comment]
@ptupitsyn Can this pr being merged? Thanks :)

**Comment by ptupitsyn** [comment]
> @ptupitsyn Can this pr being merged? Thanks :)

Yes, once we have a good TeamCity run. I'll deal with this, don't worry.

**Pull request merged**